src/engine.cc

```diff
@@ -164,16 +164,15 @@ void Engine::EnsureSearchStopped() {
 void Engine::UpdateBackendConfig() {
   const std::string backend_name =
       options_.Get<std::string>(SharedBackendParams::kBackendId);
-  const size_t cache_size =
-      options_.Get<int>(SharedBackendParams::kNNCacheSizeId);
   if (!backend_ || backend_name != backend_name_ ||
       backend_->UpdateConfiguration(options_) == Backend::NEED_RESTART) {
     backend_name_ = backend_name;
     backend_ = CreateMemCache(BackendManager::Get()->CreateFromParams(options_),
-                              cache_size);
+                              options_);
     search_->SetBackend(backend_.get());
   } else {
-    backend_->SetCacheSize(cache_size);
+    backend_->SetCacheSize(
```

Comment:
This (whether `MemCacheBackend` takes the cache size from the `kNNCacheSizeId` or from function arguments) changed back and for several times already. :-P

I think if it IS aware of the parameter, it should update it in `UpdateConfiguration`, and we don't need `SetCacheSize` at all.

Reply:
...and then `Engine` class may go back to storing `std::unique_ptr<Backend>` rather than `std::unique_ptr<MemCacheBackend>)`.

(and also see below that MemCacheBackend should probably have own BackendFactory)